db/skipped_sequence.go

```diff
@@ -0,0 +1,211 @@
+	getStartSeq() uint64
+	isRange() bool
+}
+
+// SkippedSequenceSlice stores the set of skipped sequences as an ordered slice of single skipped sequences
+// or skipped sequence ranges
+type SkippedSequenceSlice struct {
+	list []SkippedSequenceListInterface
+	lock sync.RWMutex
+}
+
+var _ SkippedSequenceListInterface = &SingleSkippedSequence{}
+
+// SingleSkippedSequence contains a single skipped sequence value + unix timestamp of the time it's created
+type SingleSkippedSequence struct {
+	seq       uint64
+	timestamp int64
+}
+
+func NewSkippedSequenceSlice() *SkippedSequenceSlice {
+	return &SkippedSequenceSlice{
+		list: []SkippedSequenceListInterface{},
+	}
+}
+
+// NewSingleSkippedSequenceEntry returns a SingleSkippedSequence with the specified sequence and the current
+// time in unix time
+func NewSingleSkippedSequenceEntry(sequence uint64) *SingleSkippedSequence {
+	return &SingleSkippedSequence{
+		seq:       sequence,
+		timestamp: time.Now().Unix(),
+	}
+}
+
+// getTimestamp returns the timestamp of the entry
+func (s *SingleSkippedSequence) getTimestamp() int64 {
+	return s.timestamp
+}
+
+// getLastSeq gets the last sequence in the range on the entry, for single items the sequence will be returned
+func (s *SingleSkippedSequence) getLastSeq() uint64 {
+	return s.seq
+}
+
+// setLastSeq sets the last sequence in the range on the entry, for single items its a no-op
+func (s *SingleSkippedSequence) setLastSeq(seq uint64) {
+	// no-op
+}
+
+// getStartSeq gets the start sequence in the range on the entry, for single items the sequence will be returned
+func (s *SingleSkippedSequence) getStartSeq() uint64 {
+	return s.seq
+}
+
+// isRange returns true if the entry is a sequence range entry, false if not
+func (s *SingleSkippedSequence) isRange() bool {
+	return false
+}
+
+// IsSkipped returns true if a given sequence exists in the skipped sequence slice
+func (s *SkippedSequenceSlice) IsSkipped(x uint64) bool {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	_, found := s.findSequence(x)
+	return found
+}
+
+// SkippedSequenceCompact will compact the entries with timestamp old enough. It will also clip
+// the capacity of the slice to length + 100 if the current capacity is 2.5x the length
+func (s *SkippedSequenceSlice) SkippedSequenceCompact(ctx context.Context, maxWait int64) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	timeNow := time.Now().Unix()
+	indexToDelete := -1
+	for _, v := range s.list {
+		timeStamp := v.getTimestamp()
+		if (timeNow - timeStamp) >= maxWait {
+			indexToDelete++
+		} else {
+			// exit early to avoid iterating through whole slice
+			break
+		}
+	}
+	// if we have incremented the index counter, we need to the index range from the slice
+	if indexToDelete > -1 {
+		s.list = slices.Delete(s.list, 0, indexToDelete+1)
+	}
+	// resize slice to reclaim memory if we need to
+	s.clip(ctx)
+}
+
+// clip will clip the capacity of the slice to a constant value if the current capacity of the slice in 2.5x the length
```

Comment:
I think this comment could clarify that the 'constant value' for the updated capacity will be the current length plus the headroom.